ipam.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (C) 2017 Nippon Telegraph and Telephone Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+
+	etcd "github.com/coreos/etcd/client"
+	"github.com/osrg/gobgp/table"
+	"golang.org/x/net/context"
+)
+
+type ipPool struct {
+	CIDR string `json:"cidr"`
+	IPIP string `json:"ipip"`
+}
+
+type ipamCache struct {
+	mu sync.RWMutex
+	m  map[string]*ipPool
+}
+
+func (c *ipamCache) add(p *ipPool, del bool) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	q, ok := c.m[p.CIDR]
+
+	if del {
+		if !ok {
+			return false
+		}
+		delete(c.m, p.CIDR)
+		return true
+	}
+
+	if ok {
+		if p.CIDR == q.CIDR && p.IPIP == q.IPIP {
+			return false
+		} else {
+			c.m[p.CIDR] = p
+			return true
+		}
+	}
+	c.m[p.CIDR] = p
+	return false
+}
+
+func (c *ipamCache) get(prefix string) *ipPool {
```

Comment:
Is this function used?  If so, does it need to handle the prefix not being in the map?